gpustack/worker/backends/sglang.py

```diff
@@ -0,0 +1,287 @@
+
+        # Set host and port
+        arguments.extend(
+            [
+                "--host",
+                "0.0.0.0",
+                "--port",
+                str(port),
+            ]
+        )
+
+        return arguments
+
+    def _get_multinode_arguments(self) -> List[str]:
+        """
+        Get multi-node deployment arguments for SGLang.
+        """
+        arguments = []
+
+        # Check if this is a multi-node deployment
+        if not (
+            hasattr(self._model_instance, 'distributed_servers')
+            and self._model_instance.distributed_servers
+            and self._model_instance.distributed_servers.subordinate_workers
+        ):
+            return []
+
+        subordinate_workers = (
+            self._model_instance.distributed_servers.subordinate_workers
+        )
+        total_nodes = len(subordinate_workers) + 1  # +1 for the current node
+
+        # Find the current node's rank
+        current_worker_ip = self._worker.ip
+        node_rank = 0  # Default to 0 (master node)
+        is_main_worker = current_worker_ip == self._model_instance.worker_ip
+
+        # Determine node rank based on worker IP
+        if not is_main_worker:
+            for idx, worker in enumerate(subordinate_workers):
+                if worker.worker_ip == current_worker_ip:
+                    node_rank = idx + 1  # Subordinate workers start from rank 1
+                    break
+
+        dist_port_range = getattr(self._config, "distributed_worker_port_range", None)
+        dist_init_port = get_free_port(port_range=dist_port_range)
+
+        # Add multi-node parameters
+        arguments.extend(
+            [
+                "--nnodes",
+                str(total_nodes),
+                "--node-rank",
+                str(node_rank),
+                "--dist-init-addr",
+                f"{self._model_instance.worker_ip}:{dist_init_port}",
+            ]
+        )
+
+        return arguments
+
+    def set_sglang_distributed_env(self, env: Dict[str, str]):
+        """
+        Set up distributed environment variables for SGLang.
+        """
+        # Set up distributed training environment
+        env["NCCL_DEBUG"] = "INFO"
+        env["NCCL_SOCKET_IFNAME"] = "^lo,docker0"
+
+        # Set master address and port for distributed training
+        if (
+            hasattr(self._model_instance, 'distributed_servers')
+            and self._model_instance.distributed_servers
+        ):
+            subordinate_workers = (
+                self._model_instance.distributed_servers.subordinate_workers
+            )
+            if subordinate_workers:
+                master_worker = subordinate_workers[0]
+                env["MASTER_ADDR"] = master_worker.worker_ip or "localhost"
+                dist_port_range = getattr(
+                    self._config, "distributed_worker_port_range", None
+                )
+                env["MASTER_PORT"] = str(get_free_port(port_range=dist_port_range))
```

Comment:
1. what does MASTER_ADDR and MASTER_PORT use for?
2. why do we choose the first suboridinate worker as MASTER?